### PR TITLE
Translate PPL-builtin functions to Spark-builtin functions

### PR DESCRIPTION
**Pull request by LantaoJin**
### Description
Add a `BuiltinFunctionTranslator` to translate PPL builtin functions to Spark builtin functions in Parser.

This PR only allows the PPL-builtin functions to be translated. In another words, using non-PPL-builtin functions in PPL query will throw an exception, even they are native Spark-builtin functions. In order to support more functions in PPL, we should update the PPL builtin functions later.
For example, following PPL builtin functions will be translated to Spark builtin functions:

String functions:

> concat
> length
> like
> lower
> ltrim
> position
> rtrim
> substring
> trim
> upper
> replace

Math Functions:

> ceil
> floor
> sqrt
> ln
> pow
> mod
> abs

Arithmetic Operator:

> \+
> \-
> \*
> \/
> %

Time Functions:

> from_unixtime
> unix_timestamp

### Issues Resolved
Resolves https://github.com/opensearch-project/opensearch-spark/issues/419, https://github.com/opensearch-project/opensearch-spark/issues/420, https://github.com/opensearch-project/opensearch-spark/issues/422, https://github.com/opensearch-project/opensearch-spark/issues/423, https://github.com/opensearch-project/opensearch-spark/issues/424, https://github.com/opensearch-project/opensearch-spark/issues/425, https://github.com/opensearch-project/opensearch-spark/issues/426, https://github.com/opensearch-project/opensearch-spark/issues/427, https://github.com/opensearch-project/opensearch-spark/issues/428, https://github.com/opensearch-project/opensearch-spark/issues/491, https://github.com/opensearch-project/opensearch-spark/issues/492, https://github.com/opensearch-project/opensearch-spark/issues/480, https://github.com/opensearch-project/opensearch-spark/issues/481, https://github.com/opensearch-project/opensearch-spark/issues/469, https://github.com/opensearch-project/opensearch-spark/issues/466, https://github.com/opensearch-project/opensearch-spark/issues/459, https://github.com/opensearch-project/opensearch-spark/issues/457, https://github.com/opensearch-project/opensearch-spark/issues/456, https://github.com/opensearch-project/opensearch-spark/issues/455

### Check List
- [-] Updated documentation
- [x] Implemented unit tests
- [x] Implemented tests for combination with other commands
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4 [hunk 1]
```diff
@@ -263,17 +263,20 @@ logicalExpression
 
 comparisonExpression
    : left = valueExpression comparisonOperator right = valueExpression  # compareExpr
+   | valueExpression IN valueList                                       # inExpr
```
Comment (LantaoJin) [inline ppl-spark-integration/src/main/antlr4/OpenSearchPPLParser.g4:266]: Added `inExpr` back since it is required by `position` functions.

**Review by YANG-DB — COMMENTED** [review]
please add additional documentation (markdown reference page) for each new command

File: ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/LogicalPlanTestUtils.scala [hunk 2]
```diff
@@ -5,13 +5,14 @@
...
 
+import org.apache.spark.sql.catalyst.analysis.AnalysisTest
 import org.apache.spark.sql.catalyst.expressions.{Alias, ExprId}
 import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan, Project}
 
 /**
  * general utility functions for ppl to spark transformation test
  */
-trait LogicalPlanTestUtils {
+trait LogicalPlanTestUtils extends AnalysisTest {
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/LogicalPlanTestUtils.scala:15]: Oops, I tried to `checkAnalysis` in unit tests I added but failed with table not found. It never used, will remove it.

**Comment by LantaoJin** [comment]
> please add additional documentation (markdown reference page) for each new command

Which file or where?

File: ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanMathFunctionsTranslatorTestSuite.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    val filterExpr = EqualTo(
+      UnresolvedAttribute("a"),
+      UnresolvedFunction("sqrt", seq(UnresolvedAttribute("b")), isDistinct = false))
+    val filterPlan = Filter(filterExpr, table)
+    val projectList = Seq(UnresolvedStar(None))
+    val expectedPlan = Project(projectList, filterPlan)
+    assertEquals(expectedPlan, logPlan)
+  }
+
+  test("test arithmetic: + - * / %") {
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanMathFunctionsTranslatorTestSuite.scala:127]: done

File: ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanMathFunctionsTranslatorTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import org.opensearch.sql.ppl.{CatalystPlanContext, CatalystQueryPlanVisitor}
+import org.opensearch.sql.ppl.utils.DataTypeTransformer.seq
+import org.scalatest.matchers.should.Matchers
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.analysis.{UnresolvedAttribute, UnresolvedFunction, UnresolvedRelation, UnresolvedStar}
+import org.apache.spark.sql.catalyst.expressions.{EqualTo, Literal}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, Project}
+
+class PPLLogicalPlanMathFunctionsTranslatorTestSuite
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanMathFunctionsTranslatorTestSuite.scala:20]: done

File: ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanStringFunctionsTranslatorTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+import org.opensearch.sql.ppl.{CatalystPlanContext, CatalystQueryPlanVisitor}
+import org.opensearch.sql.ppl.utils.DataTypeTransformer.seq
+import org.scalatest.matchers.should.Matchers
+
+import org.apache.spark.SparkFunSuite
+import org.apache.spark.sql.catalyst.analysis.{UnresolvedAttribute, UnresolvedFunction, UnresolvedRelation, UnresolvedStar}
+import org.apache.spark.sql.catalyst.expressions.{EqualTo, Like, Literal, NamedExpression}
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, Project}
+
+class PPLLogicalPlanStringFunctionsTranslatorTestSuite
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanStringFunctionsTranslatorTestSuite.scala:20]: done

**Comment by LantaoJin** [comment]
Part of expressions removed by https://github.com/opensearch-project/opensearch-spark/pull/478 are added back in order to support builtin functions translation.

**Review by penghuo — COMMENTED** [review]
High level question. What is the relationship of OpenSearch PPL and PPL in Spark? How do we make sure the syntax is not diveraged from OpenSearch PPL? Does function specification should consistent?

**Comment by LantaoJin** [comment]
> High level question. What is the relationship between OpenSearch PPL and Spark PPL? 

There is no "Spark PPL" at all as far as I'm concerned, the source of truth of PPL should be "OpenSearch PPL". I don't know why opensearch-spark copied out the PPL syntax from SQL repository. From long-term perspective, they should be generated by an unity grammar code.

> How do we ensure the syntax remains consistent with OpenSearch PPL? Should the function specifications be aligned?

This PR only translates the PPL-builtin functions to Spark-builtin functions, so required by opensearch-spark repository only. It won't introduce any split-syntax between SQL and openseach-spark repositories as long as the builtin functions in two repositories are same. If we want to add new builtin functions for any reason, for short-term, submitting PRs to this repository as well as SQL's.

**Comment by penghuo** [comment]
> From long-term perspective, they should be generated by an unity grammar code.

+1, we should do it asap. otherwise the grammer will diverage.

**Review by penghuo — APPROVED** [review]

**Review by YANG-DB — APPROVED** [review]
great work - thanks !!

**Comment by YANG-DB** [comment]
@penghuo As part of the suggested original PPL spark Issue was to extract the grammer away from the SQL plugin into an independed  repository which will become the single point of truth for the PPL syntax, documentation, grammar and such

The [issue](https://github.com/opensearch-project/sql/issues/2821)

**Comment by dai-chen** [comment]
> @penghuo As part of the suggested original PPL spark Issue was to extract the grammer away from the SQL plugin into an independed repository which will become the single point of truth for the PPL syntax, documentation, grammar and such
> 
> The [issue](https://github.com/opensearch-project/sql/issues/2821)

FYI, I think this repo is for that purpose: https://github.com/opensearch-project/piped-processing-language

**Comment by LantaoJin** [comment]
> The backport to `0.5` failed:
> 
> Then, create a pull request where the `base` branch is `0.5` and the `compare`/`head` branch is `backport/backport-448-to-0.5`.

@YANG-DB This PR is already in the branch `0.5` before you adding the label `backport 0.5`. This label is not necessary for the PRs being merged in `main` before the branch `0.5` was created from main.

**Pull request merged**